backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/scheduling/utils/DedicatedCpuPinningHelper.java

```diff
@@ -0,0 +1,212 @@
+    }
+
+    /**
+     * This function will tell if the host is capable to run a given dedicated CPU policy VM.
+     *
+     * The function first will pin the CPU pending resources to the host, then:
+     * 1. If the VM number of vCPUs is smaller or equal to the pCPU exists on a single socket:
+     *    Pass on each socket and try to allocate full cores.
+     *    If there are more vCPUs to use and there is no fully free core, try to take pCPU from any core.
+     *    (Preferring to use cores as a whole).
+     *    When there are no more vCPUs to allocate - return the list of VdsCpuUnit we chose.
+     * 2. If the VM number of vCPUs is bigger to the pCPU exists on a single socket:
+     *    - We must use multiple sockets.
+     *    Pass on each socket and try to allocate full cores (regardless being in multiple sockets).
+     *    If there are more vCPUs to use and there is no a fully free core, try to take pCPU from any core.
+     *    (Preferring to use cores as a whole).
+     *    When there are no more vCPUs to allocate - return the list of VdsCpuUnit we chose.
+     *
+     * @param vmToPendingPinnings Map of VDS GUID keys to list of VdsCpuUnits pending to be taken.
+     * @param vm VM object.
+     * @param host VDS object.
+     * @return List<VdsCpuUnit>. The list of VdsCpuUnit we are going to use. If not possible, return an empty List.
+     */
+    public synchronized List<VdsCpuUnit> allocateDedicatedCpus(VM vm, Map<Guid, List<VdsCpuUnit>> vmToPendingPinnings, VDS host) {
+        if (vm.getCpuPinningPolicy() == CpuPinningPolicy.NONE) {
+            return new ArrayList<>();
+        }
+        List<VdsCpuUnit> cpuTopology = clone(resourceManager.getVdsManager(host.getId()).getCpuTopology());
+
+        pinPendingExclusiveCpus(cpuTopology, vmToPendingPinnings, vm.getCpuPinningPolicy());
+        List<VdsCpuUnit> cpusToBeAllocated = new ArrayList<>();
+
+        if (vm.getCpuPinningPolicy() != CpuPinningPolicy.DEDICATED) {
+            String cpuPinning = vm.getCpuPinningPolicy() == CpuPinningPolicy.MANUAL ? vm.getCpuPinning() : vm.getCurrentCpuPinning();
+            Set<Integer> requestedCpus = CpuPinningHelper.getAllPinnedPCpus(cpuPinning);
+            for (Integer cpuId : requestedCpus) {
+                VdsCpuUnit vdsCpuUnit = getCpu(cpuTopology, cpuId);
+                vdsCpuUnit.setCpuPinningPolicy(vm.getCpuPinningPolicy());
+                vdsCpuUnit.setVmId(vm.getId());
+                cpusToBeAllocated.add(vdsCpuUnit);
+            }
+            return cpusToBeAllocated;
+        }
+
+
+        int cpus = vm.getNumOfCpus();
+        int cpusLeft = cpus;
+        int hostCoresPerSocket = host.getCpuCores() / host.getCpuSockets();
+        int hostThreadsPerCore = host.getCpuThreads() / host.getCpuCores();
+
+        for (int socket = 0; socket < host.getCpuSockets(); socket++) {
+            if (hostCoresPerSocket * hostThreadsPerCore >= cpus && getFreeCpusInSocket(cpuTopology, socket).size() < cpus) {
```

Comment:
> Does it mean that if the host's `hostCoresPerSocket * hostThreadsPerCore` contains enough cpus to accomodata all VMs cpus at once (in one socket), but there are not actually enough free cpus in any of them, that the pinning fails?
> 
Yes, it checks if there are enough pCPUs in one socket -> this VM must be only within single socket, we don't allow split sockets when we have such hardware. The second condition checks it for the socket. We loop here through each socket of the host until we find one that fits, otherwise we can't use that host.

> If so, I think dividing the for loop into 2 might solve that. First `for` loop would try to find a fit into one socket. If that is not successful, the second for loop would try to find a compromised pinning across sockets. Looking at the code below, I think the same applies to trying to find the whole cores - try to find the "whole" core first and only then to divide cpus on different cores.

We have hard limitation to prevent cross-sockets pinning when the hardware is enough for fit into single socket.